### PR TITLE
Add Prettier configuration & documentation

### PR DESCRIPTION
**Pull request by thibaudcolas**
Addresses #6059. This PR includes the whole of the Prettier setup except for the actual formatting, which is in #7912. I split this in two so the configuration & documentation is easier to review. Once this is reviewed, we should merge this PR and #7912 at the same time (or close this PR and merge #7912, which includes all commits).

This includes:

- The Prettier configuration, determined with the UI team (https://github.com/wagtail/wagtail/wiki/UI-team#2022-01-06-utc-evening). Prettier reads the `.editorconfig` so I’ve updated it according to what we had discussed.
- Running Prettier in CircleCI’s `frontend` job (currently not erroring, to be turned on in #7912).
- Documentation updates for contributors to know about this and use the auto-formatting
- A hook in our `.pre-commit-config.yaml` configuration
- A file to list commits which could/should be ignored in `git blame`. This is [documented for Black](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html?highlight=ignore%20revs#avoiding-ruining-git-blame). The file I copied is [from Prettier’s repository](https://github.com/prettier/prettier/blob/main/.git-blame-ignore-revs).

Prettier is set up to format all file types that are supported, with the exception of Markdown, since that can often be annoying when editing documentation without all tooling set up (as is the case when making changes via GitHub).

---

## Rebasing PRs

Here is the methodology I would recommend – to be edited with the correct commit identifiers once this is merged:

```bash
# Make a copy of your branch
git branch save/my-existing-branch my-existing-branch
# Make sure you have the latest `main` from Wagtail
git checkout main
git pull upstream main
# Rebase your work onto the version of `main` right before
# we reformatted the whole project with Prettier –
# you can then resolve conflicts that would have been present anyway.
git rebase one-commit-before-reformatting
# Install Prettier and other Wagtail development dependencies (make sure you use the correct version of Node)
npm install 
# Test-run Prettier – this should show formatting errors.
npm run lint:format
# Rebase again, this time with the reformatting as the base,
# always preserving your changes in case of conflicts, and automatically reformatting with Prettier.
git rebase \
  --strategy-option=theirs \
  --exec '(npm run format || true) && git add . && git commit --amend --no-edit --no-verify' \
  the-commit-after-reformatting
# Test-run Prettier – this should show no formatting errors.
npm run lint:format
# Finally rebase onto the latest version from Wagtail main, as per usual.
git rebase main
```

---

* [x] Do the tests still pass? (https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
* [x] Does the code comply with the style guide? (Run `make lint` from the Wagtail root)
* ~~For Python changes: Have you added tests to cover the new/fixed behaviour?~~
* ~~For front-end changes: Did you test on all of [Wagtail’s supported environments](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)?~~
* [x] For new features: Has the documentation been updated accordingly?

### CODE REVIEW

File: prettier.config.js
Comment (lb-) [inline prettier.config.js:17]: minor nitpick, can this be alpha sorted? or maybe sorted by the options in the docs (if alpha sorting does not make sense)?

File: prettier.config.js
Comment (lb-) [inline prettier.config.js:10]: will this cause any non-compiled JS to break?
I am all for this setting but just wanted to make sure we do not cause issues with JS that does not pass through a transpiler.

https://prettier.io/docs/en/options.html#trailing-commas

File: .prettierignore
Comment (lb-) [inline .prettierignore:16]: I cannot find this file in the codebase, I might be missing something but maybe this can be removed (from eslintignore also).

File: .prettierignore
Comment (lb-) [inline .prettierignore:15]: Is it worth creating a new issue to find a better way to do these inline panel JS injects?

* either by adding the script to the HTML file
* or moving the init data to data attributes (and just a small init script is needed only)

This would avoid the need for these files to be in eslint ignore also

**Review by lb- — CHANGES_REQUESTED** [review]
@thibaudcolas  looking really great.

A few minor questions and nitpicky things.

However, I wanted to ask if we needed to consider any of these rules conflicting with eslint?

I have run `npm run fix:js` and then `npm run format` in that order and then the reverse order and there are no issues but just checking.

I have run this code locally, tried to break it and tested a few things out in VSCode, this is looking great.

As for the merging of the branches/commits in your main description - I might leave that to you once the final approves/refinements come through if that is ok :)

File: prettier.config.js
Comment (thibaudcolas) [reply prettier.config.js:10]: This was a concern while we were supporting IE11, but now that we aren’t, all our target browsers will support this syntax (see the row at https://kangax.github.io/compat-table/es2016plus/#test-trailing_commas_in_function_syntax).

File: .prettierignore
Comment (thibaudcolas) [reply .prettierignore:16]: Good spot – I had copied those straight from `.eslintignore`.

File: .prettierignore
Comment (thibaudcolas) [reply .prettierignore:15]: Yep!

**Comment by thibaudcolas** [comment]
Thank you for the review @lb-, all updated with your feedback. 

> However, I wanted to ask if we needed to consider any of these rules conflicting with eslint?

Part of the reason I took such a long time to get the Prettier setup done compared to my initial plans was to get #7876 in, which included upgrading our ESLint and Stylelint configurations to versions where all stylistic rules have been disabled. So – if there are incompatibilities between the tools, they should be rare. In the case of Stylelint, it’s worth additionally mentioning that in part due to the advent of Prettier they’ve recently decided to completely stop doing stylistic linting (https://github.com/stylelint/stylelint/issues/5674).

File: .prettierignore
Comment (lb-) [reply .prettierignore:15]: @thibaudcolas here are two potential approaches

1. Just use a bit of comment syntax and string parsing to 'hide' template `{{` usage
* https://github.com/lb-/wagtail/commit/486b52d4ac48800ce153da1a618dda8fa0d9979f

2. Move most template based data to data-attributes on the `.multiple` ul in HTML and just make the two init scripts read that data, the original selector though still needs to use the template syntax
* https://github.com/lb-/wagtail/commit/c406b5b059f57c17acce4ec5509da45e8f8429e8
* This still needs to have the two JS files, but it looks possible to abstract this a bit further by moving the DOM element reading to the actual `InlinePanel` function (passing a queryselector string or a DOM element as the param instead of `opts` and handling both cases in the InlinePanel code.

Feel free to ignore for now or use if helpful, otherwise we can just address as a separate new issue.

**Review by lb- — APPROVED** [review]
@thibaudcolas all of this looks good from my perspective, not sure if we wanted to cross check this with anyone else in the core team but I am super happy to see this get in.

However, as noted - due to the git wrangling, I would rather leave the push to you.

The inline_panel js & search promotions formset js comments are not a blocker and can be resolved separately.

**Pull request merged**
